tests/unit/test_grub.py

```diff
@@ -0,0 +1,488 @@
+from unittest import mock
+
+import charms.operator_libs_linux.v0.grub as grub
+import pytest
+
+GRUB_CONFIG_EXAMPLE_BODY = """
+GRUB_RECORDFAIL_TIMEOUT=0
+GRUB_TIMEOUT=0
+GRUB_TERMINAL=console
+GRUB_CMDLINE_LINUX_DEFAULT="$GRUB_CMDLINE_LINUX_DEFAULT hugepagesz=1G"
+"""
+GRUB_CONFIG_EXAMPLE = f"""
+{grub.CONFIG_HEADER}
+{grub.CONFIG_DESCRIPTION.format(configs="#  /tmp/test-path")}
+{GRUB_CONFIG_EXAMPLE_BODY}
+"""
+EXP_GRUB_CONFIG = {
+    "GRUB_CMDLINE_LINUX_DEFAULT": "'\"$GRUB_CMDLINE_LINUX_DEFAULT hugepagesz=1G\"'",
+    "GRUB_RECORDFAIL_TIMEOUT": "0",
+    "GRUB_TERMINAL": "console",
+    "GRUB_TIMEOUT": "0",
+}
+
+
+def test_validation_error():
+    """Test validation error and it's properties."""
+    exp_key, exp_message = "test", "test message"
+
+    error = grub.ValidationError(exp_key, exp_message)
+    with pytest.raises(ValueError):
+        raise error
+
+    assert error.key == exp_key
+    assert error.message == exp_message
+    assert str(error) == exp_message
+
+
+@pytest.mark.parametrize(
+    "output, exp_result",
+    [
+        (mock.MagicMock(return_value=b"lxd"), True),
+        (mock.MagicMock(side_effect=subprocess.CalledProcessError(1, [])), False),
+    ],
+)
+def test_is_container(output, exp_result):
+    """Test helper function to validate if machine is container."""
+    with mock.patch("subprocess.check_output", new=output) as mock_check_output:
+        assert grub.is_container() == exp_result
+        mock_check_output.assert_called_once_with(
+            ["/usr/bin/systemd-detect-virt", "--container"], stderr=subprocess.STDOUT
+        )
+
+
+class BaseTestGrubLib(unittest.TestCase):
+    def setUp(self) -> None:
+        tmp_dir = tempfile.TemporaryDirectory()
+        self.tmp_dir = Path(tmp_dir.name)
+        self.addCleanup(tmp_dir.cleanup)
+
+        # configured paths
+        grub.GRUB_DIRECTORY = self.tmp_dir
+
+        # change logger
+        mocked_logger = mock.patch.object(grub, "logger")
+        self.logger = mocked_logger.start()
+        self.addCleanup(mocked_logger.stop)
+
+
+class TestGrubUtils(BaseTestGrubLib):
+    def test_parse_config(self):
+        """Test parsing example GRUB config with skipping duplicated key."""
+        stream = io.StringIO(GRUB_CONFIG_EXAMPLE)
+        result = grub._parse_config(stream)
+
+        self.assertEqual(result, EXP_GRUB_CONFIG)
+
+    def test_parse_config_with_duplicity(self):
+        """Test parsing example GRUB config with skipping duplicated key."""
+        raw_config = (
+            GRUB_CONFIG_EXAMPLE + 'GRUB_CMDLINE_LINUX_DEFAULT="$GRUB_CMDLINE_LINUX_DEFAULT pti=on"'
+        )
+        stream = io.StringIO(raw_config)
+        result = grub._parse_config(stream)
+
+        self.logger.warning.assert_called_once_with(
+            "key %s is duplicated in config", "GRUB_CMDLINE_LINUX_DEFAULT"
+        )
+        self.assertEqual(
+            result["GRUB_CMDLINE_LINUX_DEFAULT"], "'\"$GRUB_CMDLINE_LINUX_DEFAULT pti=on\"'"
```

Comment:
Yeah, per the comment above about quoting -- it's double-quoting things, when it should be unquoting the value. I would expect this:

```python
 self.assertEqual(
            result["GRUB_CMDLINE_LINUX_DEFAULT"], "$GRUB_CMDLINE_LINUX_DEFAULT pti=on",
)
```